### ISSUE TITLE
unable to communicate out of Docker container running Ansible to another Docker container

### ISSUE THREAD

**Issue by jaypipes** [ansible/ansible#48423]
##### SUMMARY

Unable to communicate from one Docker container running Ansible to another Docker container.

##### ISSUE TYPE

Bug Report

##### COMPONENT NAME

Docker (I think?)

##### ANSIBLE VERSION

2.5.0

##### CONFIGURATION

See below, Ansible is installed in a Docker container.

##### OS / ENVIRONMENT

Host is RHEL7, Docker 18.06.1-ce

##### STEPS TO REPRODUCE
I did not want to install Ansible on my host VM. So, I built a Docker container that contains Ansible, using this Dockerfile:

```Dockerfile
FROM alpine:3.7
 
ENV ANSIBLE_VERSION 2.5.0
 
ENV BUILD_PACKAGES \
  bash \
  curl \
  tar \
  openssh-client \
  sshpass \
  git \
  python \
  py-boto \
  py-dateutil \
  py-httplib2 \
  py-jinja2 \
  py-paramiko \
  py-pip \
  py-yaml \
  ca-certificates
 
# If installing ansible@testing
#RUN \
#	echo "@testing http://nl.alpinelinux.org/alpine/edge/testing" >> #/etc/apk/repositories
 
RUN set -x && \
    \
    echo "==> Adding build-dependencies..."  && \
    apk --update add --virtual build-dependencies \
      gcc \
      musl-dev \
      libffi-dev \
      openssl-dev \
      python-dev && \
    \
    echo "==> Upgrading apk and system..."  && \
    apk update && apk upgrade && \
    \
    echo "==> Adding Python runtime..."  && \
    apk add --no-cache ${BUILD_PACKAGES} && \
    pip install --upgrade pip && \
    pip install python-keyczar docker-py && \
    \
    echo "==> Installing Ansible..."  && \
    pip install ansible==${ANSIBLE_VERSION} && \
    \
    echo "==> Cleaning up..."  && \
    apk del build-dependencies && \
    rm -rf /var/cache/apk/*
 
ENV ANSIBLE_GATHERING smart
ENV ANSIBLE_HOST_KEY_CHECKING false
ENV ANSIBLE_RETRY_FILES_ENABLED false
ENV ANSIBLE_ROLES_PATH /ansible/playbooks/roles
ENV ANSIBLE_SSH_PIPELINING True
ENV PYTHONPATH /ansible/lib
ENV PATH /ansible/bin:$PATH
ENV ANSIBLE_LIBRARY /ansible/library
 
WORKDIR /ansible/playbooks
 
ENTRYPOINT ["ansible-playbook"]
```

I then started three simple Docker containers with nothing other than the base alpine 3.7 image:

```bash
for name in api db mq; do docker container run --name $name -it -d alpine:latest; done 
```

After this is done, I've got three little containers called "api", "db" and "mq". They get the IP addresses `172.17.0.2`, `172.17.0.3`, and `172.17.0.4` respectively.

Now, I put together an inventory file that looks like this:
```yaml
all:
  children:
    api:
      children:
        aio-dev.api:
    db:
      children:
        aio-dev.db:
    mq:
      children:
        aio-dev.mq:
    aio-dev.api:
      hosts:
        172.17.0.3:
          ansible_connection: docker
    aio-dev.db:
      hosts:
        172.17.0.2:
          ansible_connection: docker
    aio-dev.mq:
      hosts:
        172.17.0.4:
          ansible_connection: docker
```

I create a simple playbook that should ping each host using the Python Ansible ping module:

```yaml
---
- hosts: all
  tasks:
    - name: python ping
      ping:
```

##### EXPECTED RESULTS

I would expect to see errors about Ansible not being able to execute Python on the simple target host containers (because they don't have Python installed...).

##### ACTUAL RESULTS

However, when I attempt to execute this playbook from Ansible running inside the Docker container, I keep getting failures:

```
$ ./tools/ansible-playbook.sh playbooks/ping.yaml -i sites/aio-dev/inventory.yaml

PLAY [all] *****************************************************************************************************************************************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *****************************************************************************************************************************************************************************************************************************************************************************************************************************************************
fatal: [172.17.0.2]: FAILED! => {"msg": "docker command not found in PATH"}
fatal: [172.17.0.3]: FAILED! => {"msg": "docker command not found in PATH"}
fatal: [172.17.0.4]: FAILED! => {"msg": "docker command not found in PATH"}

PLAY RECAP *****************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
172.17.0.2                 : ok=0    changed=0    unreachable=0    failed=1
172.17.0.3                 : ok=0    changed=0    unreachable=0    failed=1
172.17.0.4                 : ok=0    changed=0    unreachable=0    failed=1
```

Looking in my Docker container with Ansible installed, I see no `docker` executable, but I *do* see the `docker-py` Python module installed correctly:

```
$ docker run -it --entrypoint /bin/sh iaas/ansible:latest
/ # which docker
/ # pip freeze | grep docker-py
docker-py==1.10.6
docker-pycreds==0.3.0
```

Can someone point me in the right direction here or tell me what I'm doing terribly wrong? Thanks very much in advance!

-jay

**Comment by sivel** [comment]
List Information
================

Hi!

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

This appears to be a user question, and we'd like to direct these kinds of things to either the mailing list or the IRC channel.

   * IRC: #ansible on irc.freenode.net   
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

If you can stop by there, we'd appreciate it.  This allows us to keep the issue tracker for bugs, pull requests, RFEs and the like.

Thank you once again and we look forward to seeing you on the list or IRC.  Thanks!

**Comment by jaypipes** [comment]
wow, that was closed pretty quickly... :(

**Comment by jaypipes** [comment]
For anyone looking to do this, the solution is to bind-mount the docker executable and docker socket when starting up the Docker container that runs Ansible:

```
docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
  -v /usr/bin/docker:/usr/bin/docker
 $ansible_container_name
```

Note that `docker-py` actually isn't necessary to install in the Docker container (contrary to information found online)

The docker connection module doesn't do connection or protocol-level communication with Docker. Instead, it shells out to the `docker` executable which means that the `docker` executable needs to be available (and its UNIX socket) in the container that runs Ansible.

**Comment by jaypipes** [comment]
Following up on this further, just in case that poor soul who is attempting to get Ansible to work in a Docker container with other Docker containers on the host happens to stumble in here...

DO NOT DO IT.

Do not install Ansible in a Docker container and try to get that Ansible to communicate with other Docker containers running on the Docker host. It just doesn't work. If you try bind-mounting /var/run/docker.sock and /usr/bin/docker into the Ansible container, you'll get issues with missing symbols because Docker isn't statically built and you'll undoubtedly run into issues with missing .so's in the Ansible container. Furrthermore, Docker-in-Docker (dind) won't work properly either because there will be confusion over which Docker daemon to query...

In short: if you want to use Ansible to run playbooks or automation tasks against target hosts and those target hosts may be Docker containers, **just install Ansible on your host**. Don't even bother trying to install Ansible in a docker container. Save yourself the heartache and trouble!

**Comment by ghost** [comment]
@jaypipes Thanks, for this report. I was banging my head with very same problem. It just baffles me that Ansible documentation has no mention of this at all. Nothing about requirement to have `docker client` on Ansible host if one wants to use `ansible_connection=docker` in [Non-SSH connection types](https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html?highlight=docker%20become#non-ssh-connection-types) section.
[Getting Started with Docker](https://docs.ansible.com/ansible/latest/scenario_guides/guide_docker.html?highlight=docker) only mentions Python packages.

But in the end I managed to achieve my goal by installing appropriate `docker client` package inside docker container and using TCP connection socket towards docker host. That of-course requires that docker host is configured accordingly to have `-H tcp://0.0.0.0:XXXX` option for docker daemon.

On CentOS 7 docker container it looked something like this:
```bash
yum install -y ansible docker-client python2-pip
pip install docker

# if docker daemon on host has "-H tcp://0.0.0.0:XXXX" then it also listens on internal docker network
export DOCKER_HOST="tcp://172.17.0.1:XXXX"
ansible -c docker -i 'container_name,' container_name -m ping
```
